### PR TITLE
Target Detection

### PR DESCRIPTION
**Pull request by benjaminruiz99**
The shaperecognition folder contains all the necessary function definitions for the purpose of recognizing shapes and determining the color of the selected contours. With this we can select targets and highlight the selected ones. Color matching code is under colorlabeler.py. Shape determination code is under  shapedetector.py

Under contour_recognition.py is the code that combines the two into a function for the purpose of recognizing targets and highlighting them and code that tests the functionality.

Tested on a Logitech webcam.

### CODE REVIEW

File: controller/contour_recognition.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import imutils
+
+vc = cv2.VideoCapture(0)
+
+if vc.isOpened():
+    rval, frame = vc.read()
+else:
+    rval = False
+
+show_frame = 1
```
Comment (abdullahelqaq) [inline controller/contour_recognition.py]: Plis remove

File: controller/contour_recognition.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+def target_recognition(image):
+    targets = list()
+    kernel = np.ones((5,5),np.uint8)
+
+    #resized = imutils.resize(image, width=300)
+    resized = image
```
Comment (abdullahelqaq) [inline controller/contour_recognition.py]: Is this really necessary?

File: controller/contour_recognition.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    gray = cv2.cvtColor(resized, cv2.COLOR_BGR2GRAY)
+    lab = cv2.cvtColor(resized, cv2.COLOR_BGR2LAB)
+    blurred = cv2.GaussianBlur(gray, (7, 7), 0)
+    #blurred = gray
```
Comment (abdullahelqaq) [inline controller/contour_recognition.py]: Can we remove these comments? They're pretty confusing

File: controller/contour_recognition.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    cv2.imshow("edges", edges)
+    cv2.imshow("dilated", dilated)
+    # gray = resized
+
+    ratio = image.shape[0] / float(edges.shape[0])
+    cnts = cv2.findContours(dilated.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
+    cnts = imutils.grab_contours(cnts)
+
+    sd = ShapeDetector()
+    cl = ColorLabeler()
```
Comment (abdullahelqaq) [inline controller/contour_recognition.py]: Can we pick better variable names please?

File: controller/contour_recognition.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    cv2.imshow("labels", image)
+    return targets
+
+while rval:
+    rval, image = vc.read()
+    target_recognition(image)
+
+    key = cv2.waitKey(20)
+    if key == 27:
+        break
```
Comment (abdullahelqaq) [inline controller/contour_recognition.py]: I imagine we can remove this

File: controller/shaperecognition/colorlabeler.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+        minDist = (np.inf, None)
+
+        for (i, row) in enumerate(self.lab):
+            d = dist.euclidean(row[0], mean)
+            if d < minDist[0]:
+                minDist = (d, i)
+
+        return self.colorNames[minDist[1]]
+
```
Comment (abdullahelqaq) [inline controller/shaperecognition/colorlabeler.py]: I think you need to untab the last line for GitHub to be happy

File: controller/shaperecognition/shapedetector.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                shape = "square" 
+            else:
+                 shape = "rectangle"
+
+        elif len(approx) == 5:
+            shape = "pentagon"
+        elif len(approx) > 5:
+            shape = "something weird"
+        else:
+            shape = "circle"
```
Comment (abdullahelqaq) [inline controller/shaperecognition/shapedetector.py]: This doesn't look like it contributes to the return value - can we remove it?

File: controller/shaperecognition/shapedetector.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        return target_info
+        
+
```
Comment (abdullahelqaq) [inline controller/shaperecognition/shapedetector.py]: See comment above about untabbing last line in file.

**Review by abdullahelqaq — CHANGES_REQUESTED** [review]
A few quick changes - otherwise looks good

**Review by therealhuy — APPROVED** [review]

**Review by abdullahelqaq — APPROVED** [review]

**Pull request merged**